BESTPRACTICES.md

```diff
@@ -4,17 +4,268 @@
+
+    - #### Message definitions
+        ```protobuf
+        message CreateOrderMsg {
+            weave.Metadata metadata = 1;
+            bytes trader = 2 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+            bytes order_book_id = 3 [(gogoproto.customname) = "OrderBookID"];
+            coin.Coin offer = 4;
+            Amount price = 5;
+        }
+        ```
+
+        As you can see above w*e define type necessary fields that will be used by `handler` to interact with application state
+
+    After defining your state and messages run `make protoc` and make sure `prototool` creates your `codec.pb.go` file successfully.
+    Now we have scaffold of our application thanks to auto-generated `codec.pb.go` file.
+
+2. #### Messages
+    > [PR#2](https://github.com/iov-one/tutorial/pull/2): _Create msgs_
+    
+    This is where we wrap our auto generated message types with useful and vital functionalities. Before getting into this section I want to remind you *validation* is **EXTREMELY**  important. First create your `msg.go` file. This is where the magic will happen :grin:
+
+    Lets bind a path to our new message
+
+    ```go
+    const (
+	    pathCreateOrderBook = "order/create_book"
+    )
+    ```
+
+    After that ensure your message is a `weave.Msg`
+
+    ```go
+    var _ weave.Msg = (*CreateOrderBookMsg)(nil)
+    ```
+
+    ```go
+    func (m CreateOrderBookMsg) Validate() error {
+	    if err := m.Metadata.Validate(); err != nil {
+		    return errors.Wrap(err, "metadata")
+	    }
+	    if err := validID(m.MarketID); err != nil {
+		    return errors.Wrap(err, "market id")
+	    }
+	    if !coin.IsCC(m.AskTicker) {
+		    return errors.Wrapf(errors.ErrCurrency, "Invalid Ask Ticker: %s", m.AskTicker)
+	    }
+	    if !coin.IsCC(m.BidTicker) {
+		    return errors.Wrapf(errors.ErrCurrency, "Invalid Bid Ticker: %s", m.BidTicker)
+	    }
+	    if m.BidTicker <= m.AskTicker {
+		    return errors.Wrapf(errors.ErrCurrency, "ask (%s) must be before bid (%s)", m.AskTicker, m.BidTicker)
+	    }
+	    return nil
+    }
+    ```
+
+    ```go
+    func validID(id []byte) error {
+	    if len(id) == 0 {
+		    return errors.Wrap(errors.ErrEmpty, "id missing")
+	    }
+	    if len(id) != 8 {
+		    return errors.Wrap(errors.ErrInput, "id is invalid length (expect 8 bytes)")
+	    }
+	    return nil
+    }
+    ```
+
+    You must have notice we even validate if ```ID```'s lenght is not 0 and equal to 8 and tickers are actually string tickers. **Remember** more validation more solid your application is.
+
+3. #### Models
+    > [PR#6](https://github.com/iov-one/tutorial/pull/6): _Create models_
+
+    We defined our state in [codec section](#codec). In order to use models in weave we have to wrap our model with some functionalities and enforce it is a **morm.Model**
+
+    Ensure our *OrderBook* fulfills **morm.Model**
+
+    ```go
+    var _ morm.Model = (*OrderBook)(nil)
+    ```
+    
+    Define **SetID** function
```

Comment:
Add that we also require a `GetID() []byte` method, but that is auto-generated if you have a protobuf field names 'ID'. If you don't want that auto-incrementing magic, don't include an `ID` field, and write Get/SetID using any other fields you deem important